tools/distrib/python/grpcio_tools/grpc_tools/main.cc

```diff
@@ -36,3 +49,137 @@ int protoc_main(int argc, char* argv[]) {
 
   return cli.Run(argc, argv);
 }
+
+namespace internal {
+
+class GeneratorContextImpl
+    : public ::google::protobuf::compiler::GeneratorContext {
+ public:
+  GeneratorContextImpl(
+      const std::vector<const ::google::protobuf::FileDescriptor*>&
+          parsed_files,
+      std::vector<std::pair<std::string, std::string>>* files_out)
+      : files_(files_out), parsed_files_(parsed_files) {}
+
+  ::google::protobuf::io::ZeroCopyOutputStream* Open(
+      const std::string& filename) {
+    files_->emplace_back(filename, "");
+    return new ::google::protobuf::io::StringOutputStream(
+        &(files_->back().second));
+  }
+
+  // NOTE(rbellevi): Equivalent to Open, since all files start out empty.
+  ::google::protobuf::io::ZeroCopyOutputStream* OpenForAppend(
+      const std::string& filename) {
+    return Open(filename);
+  }
+
+  // NOTE(rbellevi): Equivalent to Open, since all files start out empty.
+  ::google::protobuf::io::ZeroCopyOutputStream* OpenForInsert(
+      const std::string& filename, const std::string& insertion_point) {
+    return Open(filename);
+  }
+
+  void ListParsedFiles(
+      std::vector<const ::google::protobuf::FileDescriptor*>* output) {
+    *output = parsed_files_;
+  }
+
+ private:
+  std::vector<std::pair<std::string, std::string>>* files_;
+  const std::vector<const ::google::protobuf::FileDescriptor*>& parsed_files_;
+};
+
+class ErrorCollectorImpl
+    : public ::google::protobuf::compiler::MultiFileErrorCollector {
+ public:
+  ErrorCollectorImpl(std::vector<::grpc_tools::ProtocError>* errors,
+                     std::vector<::grpc_tools::ProtocWarning>* warnings)
+      : errors_(errors), warnings_(warnings) {}
+
+  void AddError(const std::string& filename, int line, int column,
+                const std::string& message) {
+    errors_->emplace_back(filename, line, column, message);
+  }
+
+  void AddWarning(const std::string& filename, int line, int column,
+                  const std::string& message) {
+    warnings_->emplace_back(filename, line, column, message);
+  }
+
+ private:
+  std::vector<::grpc_tools::ProtocError>* errors_;
+  std::vector<::grpc_tools::ProtocWarning>* warnings_;
+};
+
+static void calculate_transitive_closure(
+    const ::google::protobuf::FileDescriptor* descriptor,
+    std::vector<const ::google::protobuf::FileDescriptor*>* transitive_closure,
+    std::unordered_set<const ::google::protobuf::FileDescriptor*>* visited) {
+  for (int i = 0; i < descriptor->dependency_count(); ++i) {
+    const ::google::protobuf::FileDescriptor* dependency =
+        descriptor->dependency(i);
+    if (visited->find(dependency) == visited->end()) {
+      calculate_transitive_closure(dependency, transitive_closure, visited);
+    }
+  }
+  transitive_closure->push_back(descriptor);
+  visited->insert(descriptor);
+}
```

Comment:
Does ProtoBuf implementation provide similar function of removing duplicates?

Reply:
Discussed offline (months ago). This is not actually removing duplicates. This is an auxiliary data structure that reduces the asymptotic complexity of the overall algorithm. There *is* something similar in the command-line portion of the protoc codebase, but it's less optimal and not accessible as a programmatic API.